lib/test-collection.js

```diff
@@ -6,158 +13,168 @@ function TestCollection() {
 		throw new TypeError('Class constructor TestCollection cannot be invoked without \'new\'');
 	}
 
-	this.serial = [];
-	this.concurrent = [];
+	EventEmitter.call(this);
+
+	this.hasExclusive = false;
 	this.tests = {
+		concurrent: [],
+		serial: []
+	};
+
+	this.hooks = {
 		before: [],
 		beforeEach: [],
 		after: [],
 		afterEach: []
 	};
-	this.hasExclusive = false;
+
+	this._emitTestResult = this._emitTestResult.bind(this);
 }
 
+util.inherits(TestCollection, EventEmitter);
+
 TestCollection.prototype.add = function (test) {
 	var metadata = test.metadata;
 	var type = metadata.type;
 
 	if (!type) {
-		throw new Error('test type must be specified');
+		throw new Error('Test type must be specified');
+	}
+
+	if (!test.title && test.fn) {
+		test.title = fnName(test.fn);
+	}
+
+	// workaround for Babel giving anonymous functions a name
+	if (test.title === 'callee$0$0') {
+		test.title = null;
 	}
 
-	if (type === 'test') {
-		if (this.hasExclusive && !metadata.exclusive) {
-			return;
+	if (!test.title) {
+		if (type === 'test') {
+			test.title = '[anonymous]';
+		} else {
+			test.title = type;
 		}
+	}
 
-		if (metadata.exclusive && !this.hasExclusive) {
-			this.hasExclusive = true;
-			this.serial = [];
-			this.concurrent = [];
+	// a hook
+	if (type !== 'test') {
+		if (metadata.exclusive) {
+			throw new Error('"only" can not be used with a ' + type + ' test');
```

Comment:
Change `test` to `hook`? Perhaps `cannot` over `can not`?

`"only" cannot be used with a before hook` etc.